cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleWrapper.java

```diff
@@ -18,47 +18,160 @@
+
+  /**
+   * Get the current lifecycle of this wrapper.
+   *
+   * @return the current lifecycle of this wrapper.
+   */
+  @NotNull ModuleLifeCycle getModuleLifeCycle();
+
+  /**
+   * Get the module provider which loaded this module.
+   *
+   * @return the module provider which loaded this module.
+   */
+  @NotNull IModuleProvider getModuleProvider();
+
+  /**
+   * Get the module configuration on which base the module was created.
+   *
+   * @return the module configuration on which base the module was created.
+   */
+  @NotNull ModuleConfiguration getModuleConfiguration();
+
+  /**
+   * @deprecated Use {@link #getModuleConfiguration()} instead - same result but unwrapped.
+   */
+  @Deprecated
+  default JsonDocument getModuleConfigurationSource() {
+    return JsonDocument.newDocument(this.getModuleConfiguration());
+  }
 
+  /**
+   * Get the class loader which is responsible for this module.
+   *
+   * @return the class loader which is responsible for this module.
+   */
+  @NotNull ClassLoader getClassLoader();
+
+  /**
+   * Changes the lifecycle of this module to {@link ModuleLifeCycle#LOADED} if possible and fires all associated tasks.
+   *
+   * @return the same instance of this class, for chaining.
+   * @see #getModuleTasks()
+   * @see ModuleLifeCycle#canChangeTo(ModuleLifeCycle)
+   * @see IModuleProvider#notifyPreModuleLifecycleChange(IModuleWrapper, ModuleLifeCycle)
+   */
+  @NotNull IModuleWrapper loadModule();
+
+  /**
+   * Changes the lifecycle of this module to {@link ModuleLifeCycle#STARTED} if possible and fires all associated
+   * tasks.
+   *
+   * @return the same instance of this class, for chaining.
+   * @see #getModuleTasks()
+   * @see ModuleLifeCycle#canChangeTo(ModuleLifeCycle)
+   * @see IModuleProvider#notifyPreModuleLifecycleChange(IModuleWrapper, ModuleLifeCycle)
+   */
+  @NotNull IModuleWrapper startModule();
+
+  /**
+   * Changes the lifecycle of this module to {@link ModuleLifeCycle#STOPPED} if possible and fires all associated
+   * tasks.
+   *
+   * @return the same instance of this class, for chaining.
+   * @see #getModuleTasks()
+   * @see ModuleLifeCycle#canChangeTo(ModuleLifeCycle)
+   * @see IModuleProvider#notifyPreModuleLifecycleChange(IModuleWrapper, ModuleLifeCycle)
+   */
+  @NotNull IModuleWrapper stopModule();
+
+  /**
+   * Changes the lifecycle of this module to {@link ModuleLifeCycle#UNLOADED} if possible and fires all associated
+   * tasks. The module will be unregistered from the provider, the class loader will be closed and the state of this
+   * module changes to {@link ModuleLifeCycle#UNUSEABLE}.
+   *
+   * @return the same instance of this class, for chaining.
+   * @see #getModuleTasks()
+   * @see ModuleLifeCycle#canChangeTo(ModuleLifeCycle)
+   * @see IModuleProvider#notifyPreModuleLifecycleChange(IModuleWrapper, ModuleLifeCycle)
+   */
+  @NotNull IModuleWrapper unloadModule();
+
+  /**
+   * @deprecated Use {@link #getDataDirectory()} instead.
+   */
```

Comment:
maybe remove it now?

Reply:
I'm currently not a fan of removing these as the deprecated annotation was added in 3.4... I think we go with 3.6 to have "buffer" versions.

Reply:
If thats the case you should add @ ScheduledForRemoval?

Reply:
The provblem is that no method which return java.io.File was annotated with @ ScheduledForRemval so I'm not willing to add it here...